engine/table/src/main/java/io/deephaven/engine/table/impl/BucketedPartitionedUpdateByManager.java

```diff
@@ -0,0 +1,159 @@
+            for (UpdateByOperator op : operators) {
+                op.createInputModifiedColumnSet(source);
+                op.createOutputModifiedColumnSet(result);
+            }
+            pt = source.partitionedAggBy(List.of(), true, null, byColumns);
+        } else {
+            // no shifting will be needed, can create directly from source
+            pt = source.partitionedAggBy(List.of(), true, null, byColumns);
+
+            // create input modified column sets only
+            for (UpdateByOperator op : operators) {
+                op.createInputModifiedColumnSet(source);
+            }
+        }
+
+        // make the source->result transformer
+        transformer = source.newModifiedColumnSetTransformer(result, source.getDefinition().getColumnNamesArray());
+
+        final PartitionedTable transformed = pt.transform(t -> {
+            UpdateByBucketHelper updateBy = new UpdateByBucketHelper(
+                    description,
+                    (QueryTable) t,
+                    operators,
+                    windows,
+                    inputSources,
+                    operatorInputSourceSlots,
+                    resultSources,
+                    timestampColumnName,
+                    redirContext,
+                    control);
+
+            if (listener != null) {
+                ListenerRecorder recorder = new ListenerRecorder(description, updateBy.result, result);
+                recorder.setMergedListener(listener);
+                updateBy.result.addUpdateListener(recorder);
+
+                // add the listener only while synchronized
+                synchronized (recorders) {
+                    recorders.offerLast(recorder);
+                }
+            }
+            // add this to the bucket list
+            synchronized (buckets) {
+                buckets.offerLast(updateBy);
+            }
+            // return the table
+            return updateBy.result;
+        });
+
+        result.addParentReference(transformed);
+
+        if (source.isRefreshing()) {
+            // create a recorder instance sourced from the transformed table
+            ListenerRecorder sourceRecorder = new ListenerRecorder(description, transformed.table(), result);
+            sourceRecorder.setMergedListener(listener);
+            transformed.table().addUpdateListener(sourceRecorder);
+            result.addParentReference(listener);
+            recorders.offerLast(sourceRecorder);
+        }
+
+
+        // make a dummy update to generate the initial row keys
+        final TableUpdateImpl fakeUpdate = new TableUpdateImpl(source.getRowSet(),
+                RowSetFactory.empty(),
+                RowSetFactory.empty(),
+                RowSetShiftData.EMPTY,
+                ModifiedColumnSet.EMPTY);
+
+        // do the actual computations
+        final StateManager sm = new StateManager(fakeUpdate, true);
+        sm.processUpdate();
+    }
```

Comment:
We discussed what you really need in a simplest model.
* Source listener, to discover upstream changes, managed by `mergedListener`
* `transformed.table()` managed by `result` or `mergedListener`, with a strong reference somewhere. Maybe easier on `result` for that reason, since `addParentReference()` handles that.
* `mergedListener` to *depend* on `transformed.table()` and all constituents thereof from previous cycles, as well as the source listener that it queues a notification from.

Reply:
Really, this means you don't need a merged listener, just a regular old listener on source with extra dependencies for its notification.

Reply:
Agree, changed listener structure significantly